### ISSUE TITLE
[Feature] HTML Report: Commit Context Header/Sidebar

### ISSUE THREAD

**Issue by rwoll** [microsoft/playwright#11318]
**tl;dr:** We produce a lot of HTML Reports and as we share links to them in tickets or with teammates, it would be nice for them to have some additional context in the header of the report so we know who/what/when/etc the report was produced (most of them come from CI bots with each commit we push). If you link the report to a ticket and come back to it days/weeks later[^1], or have lots of tabs open this information becomes especially useful in understanding when/if the report is still relevant or under what conditions it was created (e.g. staging v. prod envs).

The following info would be nice to see somewhere prominently in the report (and as always, I'm happy to contribute a PR):

| Information | Info |
| ----------- | ------ |
| Start Time (Wall Time) | Generated by HTML Reporter. |
| End Time (Wall Time) | Generated by HTML Reporter. |
| Commit Author | Sourced from `PLAYWRIGHT_REPORT_ENV_COMMIT_AUTHOR` (Explicitly set, or inferred from `CI_COMMIT_AUTHOR` in GitLab, and `??` in GitHub) |
| Commit Message | Sourced from `PLAYWRIGHT_REPORT_ENV_COMMIT_MESSAGE` (Explicitly set, or inferred from `CI_COMMIT_MESSAGE` in GitLab, and `??` in GitHub) |
| Commit Hash | Sourced from `PLAYWRIGHT_REPORT_ENV_COMMIT_HASH` (Explicitly set, or inferred from `CI_COMMIT_SHA` in GitLab, and `GITHUB_SHA` in GitHub) |
| CI/CD Link | Sourced from `PLAYWRIGHT_REPORT_ENV_CI_CD_LINK` (Explicitly set, or inferred from `CI_JOB_URL` in GitLab, and `??` in GitHub) |
| Notes | Freeform notes for the user to supply via `PLAYWRIGHT_REPORT_ENV_NOTES`. |

Here's an example of what it might look like (modulo some better styling[^2]):

<img width="1055" alt="Screen Shot 2022-01-10 at 7 33 05 PM" src="https://user-images.githubusercontent.com/11915034/148877047-50ecac98-fc5a-4b42-a5f6-50a6040b46c6.png">

I picked these fields as they seem universally helpful and not unique to us. We receive test failure information via Slack notifications that also link to the (hosted) HTML Report, so the HTML Report is quickly becoming our entry point into test triaging etc.

If this is something the team/community supports, **please let me know and I will be happy to make a quick PR to the HTML Report** so the information in these variables gets inlined and presented in the UI like is done today with the the other test-time inlined report data/json.

:tada: :heart: 😄 

[^1]: We embed the timestamp and some metadata in the path to the hosted report, so we can sort of infer some minimal context now. We addionally capture context in test attachments, but that info is not discoverable when you first click on a report.

[^2]: I'd probably also make it collapsable.

**Comment by dgozman** [comment]
Note there is `TestProject.metadata` field that is somewhat related, but it is not shown in html report.

**Comment by rwoll** [comment]
> Note there is `TestProject.metadata` field that is somewhat related, but it is not shown in html report.

Thanks! If that ended up in the report UI somewhere (even just as `JSON.stringify(meta, null, "")`) that could also help us without committing Playwright to any officially supported env vars in the near term (although I'm personally a fan of some standard fields driven by env vars as a sort of best practice as I anticipate others will encounter this issue too). 

Without official env vars, but with `TestProject.metadata` being available somewhere in the UI, we'd just do something like the following in our `playwright.config.ts` (assuming GitLab):

```ts
metadata: {
  generatedAt: new Date().toString(),
  notes: process.env.CI_NOTES,
  runnerLogs: process.env.CI_JOB_URL
  commit: {
    author: process.env.CI_COMMIT_AUTHOR,
    message: process.env.CI_COMMIT_MESSAGE,
    hash: process.env.CI_COMMIT_SHA,
  },
}
```

Thinking about it more, even if some official set of report-related env vars are supported (as proposed in https://github.com/microsoft/playwright/issues/11318#issue-1098621360), it would be nice to have the metadata viewable somewhere in the UI so we/users can also include other arbitrary data that will be helpful.

As an anecdote: just this morning we were looking at two reports we received over night and were scratching our heads why they had very different behavior results: one came from testing against staging environment and the other against production environment. 🤦 I think for now, I'll also change our config for the environment to show up in the Project Name so it'll be in the UI without any of these changes landing.

**Comment by rwoll** [comment]
I'd like to implement some variant of this in the near term, but have a few open questions/thoughts.

Do we want to commit the HTML Reporter to being aware of a standard set of properties populated onto `TestProject.metadata` (e.g. `metadata.commit.message`, `metadata.notes`)? Additionally, do we want to commit to supporting a standard set of environment variables for the HTML Reporter (e.g. `PLAYWRIGHT_REPORT_ENV_COMMIT_AUTHOR`) which can be set explicitly or automagically populated from common CI/CD env vars that can drive the special metadata fields (in the case that the user doesn't want to set it via JS in the config)?

The proposed fields in https://github.com/microsoft/playwright/issues/11318#issue-1098621360 are common enough that I'd feel comfortable supporting them in a semi-structured way, but I'm also open to doing something more agnostic (e.g. just pretty-print whatever is in TestProject.metadata somewhere in the report and let users decide if/how they want to populate it from the environment).

Thoughts?
